plugins/net_plugin/net_plugin.cpp

```diff
@@ -2593,42 +2625,96 @@ namespace eosio {
       index.erase(p.first, p.second);
    }
 
-   bool dispatch_manager::add_peer_txn( const transaction_id_type& id, const time_point_sec& trx_expires,
-                                        uint32_t connection_id, const time_point_sec& now ) {
+   dispatch_manager::add_peer_txn_info dispatch_manager::add_peer_txn( const transaction_id_type& id, const time_point_sec& trx_expires, connection& c )
+   {
       fc::lock_guard g( local_txns_mtx );
-      auto tptr = local_txns.get<by_id>().find( std::make_tuple( std::ref( id ), connection_id ) );
-      bool added = (tptr == local_txns.end());
-      if( added ) {
+
+      auto& id_idx = local_txns.get<by_id>();
+      bool already_have_trx = false;
+      if (auto tptr = id_idx.find( id ); tptr != id_idx.end()) {
+         if (tptr->connection_ids.insert(c.connection_id).second)
+            ++c.trx_entries_size;
+         already_have_trx = tptr->have_trx;
+         tptr->have_trx = true;
+      } else {
          // expire at either transaction expiration or configured max expire time whichever is less
-         time_point_sec expires{now.to_time_point() + my_impl->p2p_dedup_cache_expire_time_us};
+         time_point_sec expires{fc::time_point::now() + my_impl->p2p_dedup_cache_expire_time_us};
          expires = std::min( trx_expires, expires );
          local_txns.insert( node_transaction_state{
             .id = id,
             .expires = expires,
-            .connection_id = connection_id} );
+            .connection_ids = {c.connection_id},
+            .have_trx = true } );
+         ++c.trx_entries_size;
       }
-      return added;
+      return {c.trx_entries_size, already_have_trx};
+   }
+
+   size_t dispatch_manager::add_peer_txn_notice( const transaction_id_type& id, const time_point_sec& trx_expires, connection& c )
+   {
+      fc::lock_guard g( local_txns_mtx );
+
+      auto& id_idx = local_txns.get<by_id>();
+      if (auto tptr = id_idx.find( id ); tptr != id_idx.end()) {
+         if (tptr->connection_ids.insert(c.connection_id).second)
+            ++c.trx_entries_size;
```

Comment:
Why increment the transaction count? We just received a notice.

Reply:
The count is for the entries in this multiindex. Notice or trx both add to this multiindex.

Reply:
Why do we need to keep that count?